process/process.go

```diff
@@ -0,0 +1,380 @@
+package process
+
+import (
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/renproject/hyperdrive/block"
+	"github.com/renproject/hyperdrive/id"
+)
+
+// Step in the consensus algorithm.
+type Step uint8
+
+// Define all Steps.
+const (
+	StepNil       = Step(0)
+	StepPropose   = Step(1)
+	StepPrevote   = Step(2)
+	StepPrecommit = Step(3)
+)
+
+// A Blockchain defines a storage interface for Blocks that is based around
+// Height.
+type Blockchain interface {
+	InsertBlockAtHeight(block.Height, block.Block)
+	BlockAtHeight(block.Height) (block.Block, bool)
+	BlockExistsAtHeight(block.Height) bool
+}
+
+// A Proposer builds a `block.Block` for proposals.
+type Proposer interface {
+	BlockProposal(block.Height, block.Round) block.Block
+}
+
+// A Validator validates a `block.Block` that has been proposed.
+type Validator interface {
+	IsBlockValid(block.Block) bool
+}
+
+// An Observer is notified when note-worthy events happen for the first time.
+type Observer interface {
+	DidCommitBlock(block.Height)
+}
+
+// A Scheduler determines which `id.Signatory` should be broadcasting
+// proposals in at a given `block.Height` and `block.Round`.
+type Scheduler interface {
+	Schedule(block.Height, block.Round) id.Signatory
+}
+
+// A Broadcaster sends a Message to as many Processes in the network as
+// possible.
+type Broadcaster interface {
+	Broadcast(Message)
+}
+
+// A Timer determines the timeout duration at a given Step and `block.Round`.
+type Timer interface {
+	Timeout(step Step, round block.Round) time.Duration
+}
+
+// Processes defines a wrapper type around the []Process type.
+type Processes []Process
+
+// A Process defines a state machine in the distributed replicated state
+// machine. See https://arxiv.org/pdf/1807.04938.pdf for more information.
+type Process struct {
+	mu *sync.Mutex
+
+	signatory  id.Signatory
+	blockchain Blockchain
+	state      State
+
+	proposer    Proposer
+	validator   Validator
+	scheduler   Scheduler
+	broadcaster Broadcaster
+	timer       Timer
+	observer    Observer
+}
+
+// New Process initialised to the default state, starting in the first round.
+func New(signatory id.Signatory, blockchain Blockchain, state State, proposer Proposer, validator Validator, observer Observer, broadcaster Broadcaster, scheduler Scheduler, timer Timer) Process {
+	p := Process{
+		mu: new(sync.Mutex),
+
+		signatory:  signatory,
+		blockchain: blockchain,
+		state:      state,
+
+		proposer:    proposer,
+		validator:   validator,
+		observer:    observer,
+		broadcaster: broadcaster,
+		scheduler:   scheduler,
+		timer:       timer,
+	}
+	p.StartRound(0)
```

Comment:
Actually, I think it's even more than that. We probably shouldn't call `StartRound(0)` unless the `State` has a current `Height` and `Round` of 0 (signally that we are at the genesis block). Because the point of the persistent `State` is to not have to reset our round data between reboots.

Reply:
Fixed.